### PR TITLE
Tradeable exits

### PR DESCRIPTION
**Pull request by minhptruong**
- Augmented Arbitrum protocol to enable instantaneous trade of unsettled assets. Thusly, contributing to higher market liquidity and fostering a secondary asset market for these positions.
- Users are now endowed with the capacity to retrieve information regarding the current owner of an unconfirmed payment, verify such payment exists, and trade these assets.

### CODE REVIEW

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 14
           

Complexity increasing per file
==============================
- packages/arb-validator/rollupvalidator/txtracker.go  5
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/validatorserver/server.pb.go  -1
         

Clones added
============
- packages/arb-bridge-eth/test/globalInbox.js  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5528055&bid=18117574)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 14
           

Complexity increasing per file
==============================
- packages/arb-validator/rollupvalidator/txtracker.go  5
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/validatorserver/server.pb.go  -1
         

Clones added
============
- packages/arb-bridge-eth/test/globalInbox.js  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5528055&bid=18117574)

File: packages/arb-bridge-eth/contracts/PaymentRecords.sol [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pragma solidity ^0.5.3;
+
+contract PaymentRecords {
+	mapping(bytes32 => mapping(uint256 => mapping(address => address))) paymentMap;
```
Comment (edfelten) [inline packages/arb-bridge-eth/contracts/PaymentRecords.sol:20]: Should put a comment somewhere (either here or on getPaymentOwner) making clear that the presence of a record here does not imply that a payment exists.  A record means only that *if* such a payment exists, this is who owns it.

File: packages/arb-bridge-eth/contracts/GlobalInbox.sol [hunk 2]
```diff
@@ -99,7 +118,9 @@ contract GlobalInbox is GlobalEthWallet, GlobalFTWallet, GlobalNFTWallet, IGloba
             if (!valid) {
                 return (false, startOffset);
             }
-            transferEth(msg.sender, to, value);
+
+            address paymentOwner = getPaymentOwner(to, nodeHash, messageIndex);
+            transferEth(msg.sender, paymentOwner, value);
```
Comment (edfelten) [inline packages/arb-bridge-eth/contracts/GlobalInbox.sol:123]: Can we give back the storage used for the payment record here, after the transfer succeeds?  This might save a little bit of gas.  But only do it if it's definitely safe--better to waste a little gas than risk dropping a payment.

File: packages/arb-validator/rollupvalidator/txtracker.go [hunk 3]
```diff
@@ -229,6 +247,20 @@ func (tr *txTracker) processFinalizedAssertion(assertion rollup.FinalizedAsserti
 
 func (tr *txTracker) processRequest(request validatorRequest) {
 	switch request := request.(type) {
+	case outputMsgRequest:
+		assertionVal := newAssertionInfo()
+		for _, assertion := range tr.assertionInfo {
+			if request.assertionNodeHash == assertion.AssertNodeHash {
```
Comment (hkalodner) [inline packages/arb-validator/rollupvalidator/txtracker.go:253]: It'd be good to add a separate mapping of nodeHash -> assertion so that we can do this efficiently

File: packages/arb-provider-ethers/src/lib/wallet.ts [hunk 4]
```diff
@@ -162,6 +162,23 @@ export class ArbWallet extends ethers.Signer {
...
+  ): Promise<ethers.providers.TransactionResponse> {
+    const msgIndex = ethers.utils.bigNumberify(messageIndex)
+    const globalInbox = await this.globalInboxConn()
+    const tx = await globalInbox.transferPayment(
+      originalOwner,
+      newOwner,
+      nodeHash,
+      msgIndex
+    )
+    return this.provider._wrapTransaction(tx, tx.hash)
```
Comment (hkalodner) [inline packages/arb-provider-ethers/src/lib/wallet.ts:179]: This is just a standard L1 transaction so it doesn't need to be resolved by the arbitrum provider. We can just return `tx` directly instead of wrapping it. It might be better to not have this in the wallet at all and instead just put it in the token bridge sdk.

File: packages/arb-provider-ethers/src/lib/provider.ts [hunk 5]
```diff
@@ -149,47 +154,6 @@ export class ArbProvider extends ethers.providers.BaseProvider {
     )
   }
 
-  // public async getValidatorAddresses(): Promise<string[]> {
```
Comment (hkalodner) [inline packages/arb-provider-ethers/src/lib/provider.ts:152]: Fine to delete all of this, but try to keep feature PRs isolated to just the work of adding that feature rather than including other unrelated changes.

File: packages/arb-provider-ethers/src/lib/provider.ts [hunk 6]
```diff
@@ -45,6 +45,7 @@ function sleep(ms: number): Promise<void> {
 // EthBridge event names
 const EB_EVENT_VMC = 'VMCreated'
 const EB_EVENT_CDA = 'RollupAsserted'
+const EB_EVENT_NODE = 'RollupStakeMoved'
```
Comment (hkalodner) [inline packages/arb-provider-ethers/src/lib/provider.ts:48]: Why was this added? It looks unused

File: packages/arb-provider-ethers/src/lib/provider.ts [hunk 7]
```diff
@@ -16,7 +16,7 @@
 /* eslint-env node */
 'use strict'
 
-import { ArbClient, EVMCode, EVMResult, TxMessage } from './client'
+import { ArbClient, EVMCode, EVMResult, TxMessage, ArbMessage } from './client'
```
Comment (hkalodner) [inline packages/arb-provider-ethers/src/lib/provider.ts:19]: ArbMessage was added but is never used

File: packages/arb-provider-ethers/src/lib/client.ts [hunk 8]
```diff
@@ -400,13 +405,57 @@ interface MessageResult {
   evmVal: EVMResult
 }
 
+interface OutputMessage {
+  outputMsg: ArbValue.Value
+  vmId: string
```
Comment (hkalodner) [inline packages/arb-provider-ethers/src/lib/client.ts:410]: Is there any reason the `vmId` field is included? If not it'd be cleaner to not have it

File: packages/arb-bridge-eth/contracts/PaymentRecords.sol [hunk 9]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2019-2020, Offchain Labs, Inc.
```
Comment (hkalodner) [inline packages/arb-bridge-eth/contracts/PaymentRecords.sol:2]: This should just be 2020

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 11
           

Complexity increasing per file
==============================
- packages/arb-validator/rollupvalidator/txtracker.go  4
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/validatorserver/server.pb.go  -1
         

Clones added
============
- packages/arb-bridge-eth/test/globalInbox.js  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5528055&bid=18117574)

File: packages/arb-provider-ethers/src/lib/client.ts [hunk 10]
```diff
@@ -444,12 +443,10 @@ export class ArbClient {
       }
     )
     if (msgResult.found) {
-      const vmId = await this.getVmID()
       const val = ArbValue.unmarshal(msgResult.rawVal)
 
       return {
         outputMsg: val,
-        vmId: vmId,
```
Comment (hkalodner) [inline packages/arb-provider-ethers/src/lib/client.ts]: Why was this changed?

**Comment by hkalodner** [comment]
Before merging it would be good to add a summary of the major changes included in this PR

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 11
           

Complexity increasing per file
==============================
- packages/arb-validator/rollupvalidator/txtracker.go  4
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/validatorserver/server.pb.go  -45
         

Clones added
============
- packages/arb-provider-go/arbsys.go  2
- packages/arb-bridge-eth/test/globalInbox.js  4
         

Clones removed
==============
+ packages/arb-validator-core/validatorserver/server.pb.go  -26
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5528055&bid=18117574)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 11
           

Complexity increasing per file
==============================
- packages/arb-validator/rollupvalidator/txtracker.go  4
         

Complexity decreasing per file
==============================
+ packages/arb-validator-core/validatorserver/server.pb.go  -45
         

Clones added
============
- packages/arb-provider-go/arbsys.go  2
- packages/arb-bridge-eth/test/globalInbox.js  4
         

Clones removed
==============
+ packages/arb-validator-core/validatorserver/server.pb.go  -26
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5528055&bid=18117574)
